Centaur: empty_string test fails on MySQL databases

The following stack trace appears:

```
2017-01-27 11:43:10,369 cromwell-system-akka.dispatchers.engine-dispatcher-19 ERROR - WorkflowManagerActor Workflow 268771d3-0303-45b1-ba0f-9c93c27b6784 failed (during ExecutingWorkflowState): JobStore write failure: Invalid position in SerialClob object set
java.lang.Exception: JobStore write failure: Invalid position in SerialClob object set
	at cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor$$anonfun$11.applyOrElse(EngineJobExecutionActor.scala:239)
	at cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor$$anonfun$11.applyOrElse(EngineJobExecutionActor.scala:235)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.FSM$class.processEvent(FSM.scala:663)
	at cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor.akka$actor$LoggingFSM$$super$processEvent(EngineJobExecutionActor.scala:29)
	at akka.actor.LoggingFSM$class.processEvent(FSM.scala:799)
	at cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor.processEvent(EngineJobExecutionActor.scala:29)
	at akka.actor.FSM$class.akka$actor$FSM$$processMsg(FSM.scala:657)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:651)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:496)
	at cromwell.engine.workflow.lifecycle.execution.EngineJobExecutionActor.aroundReceive(EngineJobExecutionActor.scala:29)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: javax.sql.rowset.serial.SerialException: Invalid position in SerialClob object set
	at javax.sql.rowset.serial.SerialClob.getSubString(SerialClob.java:269)
	at com.mysql.jdbc.PreparedStatement.setClob(PreparedStatement.java:3171)
	at com.zaxxer.hikari.proxy.PreparedStatementJavassistProxy.setClob(PreparedStatementJavassistProxy.java)
	at slick.driver.JdbcTypesComponent$JdbcTypes$ClobJdbcType.setValue(JdbcTypesComponent.scala:148)
	at slick.driver.JdbcTypesComponent$JdbcTypes$ClobJdbcType.setValue(JdbcTypesComponent.scala:146)
	at slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:18)
	at slick.jdbc.BaseResultConverter.set(JdbcResultConverter.scala:10)
	at slick.relational.ProductResultConverter.set(ResultConverter.scala:71)
	at slick.relational.ProductResultConverter.set(ResultConverter.scala:46)
	at slick.relational.TypeMappingResultConverter.set(ResultConverter.scala:138)
	at slick.driver.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction$$anonfun$run$8$$anonfun$apply$1.apply(JdbcActionComponent.scala:519)
	at slick.driver.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction$$anonfun$run$8$$anonfun$apply$1.apply(JdbcActionComponent.scala:517)
	at slick.jdbc.JdbcBackend$SessionDef$class.withPreparedInsertStatement(JdbcBackend.scala:354)
	at slick.jdbc.JdbcBackend$BaseSession.withPreparedInsertStatement(JdbcBackend.scala:407)
	at slick.driver.JdbcActionComponent$ReturningInsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:636)
	at slick.driver.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction$$anonfun$run$8.apply(JdbcActionComponent.scala:517)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:285)
	at slick.driver.JdbcActionComponent$InsertActionComposerImpl$MultiInsertAction.run(JdbcActionComponent.scala:522)
	at slick.driver.JdbcActionComponent$SimpleJdbcDriverAction.run(JdbcActionComponent.scala:32)
	at slick.driver.JdbcActionComponent$SimpleJdbcDriverAction.run(JdbcActionComponent.scala:29)
	at slick.backend.DatabaseComponent$DatabaseDef$$anon$2.liftedTree1$1(DatabaseComponent.scala:237)
	at slick.backend.DatabaseComponent$DatabaseDef$$anon$2.run(DatabaseComponent.scala:237)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

--- a/database/migration/src/main/resources/changelog.xml
+++ b/database/migration/src/main/resources/changelog.xml
@@ -57,6 +57,7 @@
     <include file="changesets/workflow_store_labels_file.xml" relativeToChangelogFile="true" />
     <include file="changesets/embiggen_detritus_value.xml" relativeToChangelogFile="true" />
     <include file="changesets/standardize_column_names_patches.xml" relativeToChangelogFile="true" />
+    <include file="changesets/nullable_lobs.xml" relativeToChangelogFile="true" />
 </databaseChangeLog>
 <!--
 

--- a/database/migration/src/main/resources/changesets/nullable_lobs.xml
+++ b/database/migration/src/main/resources/changesets/nullable_lobs.xml
@@ -1,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.3.xsd">
+    <changeSet author="kshakir" id="nullable_lobs">
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="CALL_CACHING_DETRITUS_ENTRY" columnName="DETRITUS_VALUE"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="CALL_CACHING_SIMPLETON_ENTRY" columnName="SIMPLETON_VALUE"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="JOB_STORE_SIMPLETON_ENTRY" columnName="SIMPLETON_VALUE"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="WORKFLOW_STORE_ENTRY" columnName="WORKFLOW_DEFINITION"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="WORKFLOW_STORE_ENTRY" columnName="WORKFLOW_INPUTS"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="WORKFLOW_STORE_ENTRY" columnName="WORKFLOW_OPTIONS"/>
+        <dropNotNullConstraint columnDataType="LONGTEXT"
+                               tableName="WORKFLOW_STORE_ENTRY" columnName="CUSTOM_LABELS"/>
+    </changeSet>
+</databaseChangeLog>

--- a/database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingDetritusEntryComponent.scala
+++ b/database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingDetritusEntryComponent.scala
@@ -16,7 +16,7 @@ trait CallCachingDetritusEntryComponent {
 
     def detritusKey = column[String]("DETRITUS_KEY")
 
-    def detritusValue = column[Clob]("DETRITUS_VALUE")
+    def detritusValue = column[Option[Clob]]("DETRITUS_VALUE")
 
     def callCachingEntryId = column[Int]("CALL_CACHING_ENTRY_ID")
 

--- a/database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingSimpletonEntryComponent.scala
+++ b/database/sql/src/main/scala/cromwell/database/slick/tables/CallCachingSimpletonEntryComponent.scala
@@ -16,7 +16,7 @@ trait CallCachingSimpletonEntryComponent {
 
     def simpletonKey = column[String]("SIMPLETON_KEY")
 
-    def simpletonValue = column[Clob]("SIMPLETON_VALUE")
+    def simpletonValue = column[Option[Clob]]("SIMPLETON_VALUE")
 
     def wdlType = column[String]("WDL_TYPE")
 

--- a/database/sql/src/main/scala/cromwell/database/slick/tables/JobStoreSimpletonEntryComponent.scala
+++ b/database/sql/src/main/scala/cromwell/database/slick/tables/JobStoreSimpletonEntryComponent.scala
@@ -16,7 +16,7 @@ trait JobStoreSimpletonEntryComponent {
 
     def simpletonKey = column[String]("SIMPLETON_KEY")
 
-    def simpletonValue = column[Clob]("SIMPLETON_VALUE")
+    def simpletonValue = column[Option[Clob]]("SIMPLETON_VALUE")
 
     def wdlType = column[String]("WDL_TYPE")
 

--- a/database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
+++ b/database/sql/src/main/scala/cromwell/database/slick/tables/WorkflowStoreEntryComponent.scala
@@ -15,13 +15,13 @@ trait WorkflowStoreEntryComponent {
 
     def workflowExecutionUuid = column[String]("WORKFLOW_EXECUTION_UUID")
 
-    def workflowDefinition = column[Clob]("WORKFLOW_DEFINITION")
+    def workflowDefinition = column[Option[Clob]]("WORKFLOW_DEFINITION")
 
-    def workflowInputs = column[Clob]("WORKFLOW_INPUTS")
+    def workflowInputs = column[Option[Clob]]("WORKFLOW_INPUTS")
 
-    def workflowOptions = column[Clob]("WORKFLOW_OPTIONS")
+    def workflowOptions = column[Option[Clob]]("WORKFLOW_OPTIONS")
 
-    def customLabels = column[Clob]("CUSTOM_LABELS")
+    def customLabels = column[Option[Clob]]("CUSTOM_LABELS")
 
     def workflowState = column[String]("WORKFLOW_STATE", O.Length(15))
 

--- a/database/sql/src/main/scala/cromwell/database/sql/SqlConverters.scala
+++ b/database/sql/src/main/scala/cromwell/database/sql/SqlConverters.scala
@@ -18,21 +18,37 @@ object SqlConverters {
     def toSystemTimestamp = Timestamp.valueOf(offsetDateTime.atZoneSameInstant(ZoneId.systemDefault).toLocalDateTime)
   }
 
-  implicit class ClobToRawString(val clob: Clob) extends AnyVal {
-    def toRawString: String = clob.getSubString(1, clob.length.toInt) // yes, it starts at 1
+  implicit class ClobOptionToRawString(val clobOption: Option[Clob]) extends AnyVal {
+    // yes, it starts at 1
+    def toRawStringOption: Option[String] = clobOption.map(clob => clob.getSubString(1, clob.length.toInt))
 
-    def parseSystemTimestamp: Timestamp = OffsetDateTime.parse(toRawString).toSystemTimestamp
+    def toRawString: String = toRawStringOption.getOrElse("")
+
+    def parseSystemTimestampOption: Option[Timestamp] = toRawStringOption map { rawString =>
+      OffsetDateTime.parse(rawString).toSystemTimestamp
+    }
   }
 
-  implicit class StringToClob(val str: String) extends AnyVal {
-    def toClob: Clob = new SerialClob(str.toCharArray)
+  implicit class StringOptionToClobOption(val strOption: Option[String]) extends AnyVal {
+    def toClob: Option[Clob] = strOption.flatMap(_.toClob)
   }
 
-  implicit class BlobToBytes(val blob: Blob) extends AnyVal {
-    def toBytes: Array[Byte] = blob.getBytes(1, blob.length.toInt) // yes, it starts at 1
+  implicit class StringToClobOption(val str: String) extends AnyVal {
+    def toClob: Option[Clob] = if (str.isEmpty) None else Option(new SerialClob(str.toCharArray))
   }
 
-  implicit class StringToBlob(val bytes: Array[Byte]) extends AnyVal {
-    def toBlob: Blob = new SerialBlob(bytes)
+  implicit class BlobToBytes(val blobOption: Option[Blob]) extends AnyVal {
+    // yes, it starts at 1
+    def toBytesOption: Option[Array[Byte]] = blobOption.map(blob => blob.getBytes(1, blob.length.toInt))
+
+    def toBytes: Array[Byte] = toBytesOption.getOrElse(Array.empty)
+  }
+
+  implicit class BytesOptionToBlob(val bytesOption: Option[Array[Byte]]) extends AnyVal {
+    def toBlob: Option[Blob] = bytesOption.flatMap(_.toBlob)
+  }
+
+  implicit class BytesToBlob(val bytes: Array[Byte]) extends AnyVal {
+    def toBlob: Option[Blob] = if (bytes.isEmpty) None else Option(new SerialBlob(bytes))
   }
 }

--- a/database/sql/src/main/scala/cromwell/database/sql/tables/CallCachingDetritusEntry.scala
+++ b/database/sql/src/main/scala/cromwell/database/sql/tables/CallCachingDetritusEntry.scala
@@ -5,7 +5,7 @@ import java.sql.Clob
 case class CallCachingDetritusEntry
 (
   detritusKey: String,
-  detritusValue: Clob,
+  detritusValue: Option[Clob],
   callCachingEntryId: Option[Int] = None,
   callCachingDetritusEntryId: Option[Int] = None
 )

--- a/database/sql/src/main/scala/cromwell/database/sql/tables/CallCachingSimpletonEntry.scala
+++ b/database/sql/src/main/scala/cromwell/database/sql/tables/CallCachingSimpletonEntry.scala
@@ -5,7 +5,7 @@ import java.sql.Clob
 case class CallCachingSimpletonEntry
 (
   simpletonKey: String,
-  simpletonValue: Clob,
+  simpletonValue: Option[Clob],
   wdlType: String,
   callCachingEntryId: Option[Int] = None,
   callCachingSimpletonEntryId: Option[Int] = None

--- a/database/sql/src/main/scala/cromwell/database/sql/tables/JobStoreSimpletonEntry.scala
+++ b/database/sql/src/main/scala/cromwell/database/sql/tables/JobStoreSimpletonEntry.scala
@@ -5,7 +5,7 @@ import java.sql.Clob
 case class JobStoreSimpletonEntry
 (
   simpletonKey: String,
-  simpletonValue: Clob,
+  simpletonValue: Option[Clob],
   wdlType: String,
   jobStoreEntryId: Option[Int] = None,
   jobStoreSimpletonEntryId: Option[Int] = None

--- a/database/sql/src/main/scala/cromwell/database/sql/tables/WorkflowStoreEntry.scala
+++ b/database/sql/src/main/scala/cromwell/database/sql/tables/WorkflowStoreEntry.scala
@@ -5,12 +5,12 @@ import java.sql.{Blob, Clob, Timestamp}
 case class WorkflowStoreEntry
 (
   workflowExecutionUuid: String,
-  workflowDefinition: Clob,
-  workflowInputs: Clob,
-  workflowOptions: Clob,
+  workflowDefinition: Option[Clob],
+  workflowInputs: Option[Clob],
+  workflowOptions: Option[Clob],
   workflowState: String,
   submissionTime: Timestamp,
   importsZip: Option[Blob],
-  customLabels: Clob,
+  customLabels: Option[Clob],
   workflowStoreEntryId: Option[Int] = None
 )

--- a/engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala
+++ b/engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala
@@ -57,7 +57,7 @@ case class SqlWorkflowStore(sqlDatabase: WorkflowStoreSqlDatabase) extends Workf
       workflowStoreEntry.workflowInputs.toRawString,
       workflowStoreEntry.workflowOptions.toRawString,
       workflowStoreEntry.customLabels.toRawString,
-      workflowStoreEntry.importsZip.map(_.toBytes)
+      workflowStoreEntry.importsZip.toBytesOption
     )
     WorkflowToStart(
       WorkflowId.fromString(workflowStoreEntry.workflowExecutionUuid),
@@ -74,7 +74,7 @@ case class SqlWorkflowStore(sqlDatabase: WorkflowStoreSqlDatabase) extends Workf
       customLabels = workflowSourceFiles.labelsJson.toClob,
       workflowState = WorkflowStoreState.Submitted.toString,
       submissionTime = OffsetDateTime.now.toSystemTimestamp,
-      importsZip = workflowSourceFiles.importsZipFileOption.map(_.toBlob)
+      importsZip = workflowSourceFiles.importsZipFileOption.toBlob
     )
   }
 

--- a/engine/src/main/scala/cromwell/jobstore/SqlJobStore.scala
+++ b/engine/src/main/scala/cromwell/jobstore/SqlJobStore.scala
@@ -49,7 +49,7 @@ class SqlJobStore(sqlDatabase: JobStoreSqlDatabase) extends JobStore {
           key.attempt,
           jobSuccessful = false,
           returnCode,
-          Option(throwable.getMessage.toClob),
+          throwable.getMessage.toClob,
           Option(retryable))
         JobStoreJoin(entry, Seq.empty)
     }
@@ -64,8 +64,9 @@ class SqlJobStore(sqlDatabase: JobStoreSqlDatabase) extends JobStore {
             val simpletons = simpletonEntries map toSimpleton
             val jobOutputs = WdlValueBuilder.toJobOutputs(taskOutputs, simpletons)
             JobResultSuccess(returnCode, jobOutputs)
-          case JobStoreEntry(_, _, _, _, false, returnCode, Some(exceptionMessage), Some(retryable), _) =>
-            JobResultFailure(returnCode, JobAlreadyFailedInJobStore(jobStoreKey.tag, exceptionMessage.toRawString),
+          case JobStoreEntry(_, _, _, _, false, returnCode, Some(_), Some(retryable), _) =>
+            JobResultFailure(returnCode,
+              JobAlreadyFailedInJobStore(jobStoreKey.tag, entry.exceptionMessage.toRawString),
               retryable)
           case bad =>
             throw new Exception(s"Invalid contents of JobStore table: $bad")

--- a/services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala
+++ b/services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala
@@ -41,12 +41,12 @@ object MetadataDatabaseAccess {
     def toSummary: WorkflowMetadataSummaryEntry = {
       val base = baseSummary(metadatum.workflowExecutionUuid)
       metadatum.metadataKey match {
-        case WorkflowMetadataKeys.Name => base.copy(workflowName = metadatum.metadataValue.map(_.toRawString))
-        case WorkflowMetadataKeys.Status => base.copy(workflowStatus = metadatum.metadataValue.map(_.toRawString))
+        case WorkflowMetadataKeys.Name => base.copy(workflowName = metadatum.metadataValue.toRawStringOption)
+        case WorkflowMetadataKeys.Status => base.copy(workflowStatus = metadatum.metadataValue.toRawStringOption)
         case WorkflowMetadataKeys.StartTime =>
-          base.copy(startTimestamp = metadatum.metadataValue.map(_.parseSystemTimestamp))
+          base.copy(startTimestamp = metadatum.metadataValue.parseSystemTimestampOption)
         case WorkflowMetadataKeys.EndTime =>
-          base.copy(endTimestamp = metadatum.metadataValue.map(_.parseSystemTimestamp))
+          base.copy(endTimestamp = metadatum.metadataValue.parseSystemTimestampOption)
       }
     }
   }
@@ -77,7 +77,7 @@ trait MetadataDatabaseAccess {
       val valueType = metadataEvent.value map { _.valueType.typeName }
       val jobKey = key.jobKey map { jk => (jk.callFqn, jk.index, jk.attempt) }
       MetadataEntry(workflowUuid, jobKey.map(_._1), jobKey.flatMap(_._2), jobKey.map(_._3),
-        key.key, value.map(_.toClob), valueType, timestamp)
+        key.key, value.toClob, valueType, timestamp)
     }
     databaseInterface.addMetadataEntries(metadata)
   }
@@ -91,10 +91,9 @@ trait MetadataDatabaseAccess {
       } yield MetadataJobKey(callFqn, m.jobIndex, attempt)
 
       val key = MetadataKey(workflowId, metadataJobKey, m.metadataKey)
-      val value =  for {
-        mValue <- m.metadataValue
-        mType <- m.metadataValueType
-      } yield MetadataValue(mValue.toRawString, MetadataType.fromString(mType))
+      val value = m.metadataValueType.map(mType =>
+        MetadataValue(m.metadataValue.toRawString, MetadataType.fromString(mType))
+      )
 
       MetadataEvent(key, value, m.metadataTimestamp.toSystemOffsetDateTime)
     }

--- a/services/src/test/scala/cromwell/services/ServicesStoreSpec.scala
+++ b/services/src/test/scala/cromwell/services/ServicesStoreSpec.scala
@@ -1,7 +1,8 @@
 package cromwell.services
 
 import java.io.{ByteArrayOutputStream, PrintStream}
-import java.sql.Connection
+import java.sql.{Connection, JDBCType}
+import java.time.OffsetDateTime
 
 import better.files._
 import com.typesafe.config.ConfigFactory
@@ -9,8 +10,9 @@ import cromwell.core.Tags._
 import cromwell.core.WorkflowId
 import cromwell.database.migration.liquibase.LiquibaseUtils
 import cromwell.database.slick.SlickDatabase
+import cromwell.database.sql.SqlConverters._
 import cromwell.database.sql.joins.JobStoreJoin
-import cromwell.database.sql.tables.JobStoreEntry
+import cromwell.database.sql.tables.{JobStoreEntry, JobStoreSimpletonEntry, WorkflowStoreEntry}
 import liquibase.diff.DiffResult
 import liquibase.diff.output.DiffOutputControl
 import liquibase.diff.output.changelog.DiffToChangeLog
@@ -153,8 +155,8 @@ class ServicesStoreSpec extends FlatSpec with Matchers with ScalaFutures with St
         val expected = s"PK_${primaryKey.table.name}"
         if (actual != expected) {
           misnamed :+=
-            s"""|PrimaryKey: $actual
-                |Should Be:  $expected
+            s"""|  PrimaryKey: $actual
+                |  Should be:  $expected
                 |""".stripMargin
         }
       }
@@ -164,8 +166,8 @@ class ServicesStoreSpec extends FlatSpec with Matchers with ScalaFutures with St
         val expected = s"FK_${foreignKey.fkTable.name}_${foreignKey.fkColumn}"
         if (actual != expected) {
           misnamed :+=
-            s"""|ForeignKey: $actual
-                |Should Be:  $expected
+            s"""|  ForeignKey: $actual
+                |  Should be:  $expected
                 |""".stripMargin
         }
       }
@@ -183,35 +185,56 @@ class ServicesStoreSpec extends FlatSpec with Matchers with ScalaFutures with St
 
           if (actual != expected) {
             misnamed :+=
-              s"""|Index:     $actual
-                  |Should Be: $expected
+              s"""|  Index:     $actual
+                  |  Should be: $expected
                   |""".stripMargin
           }
-      }
-
-      if (misnamed.nonEmpty) {
-        fail(misnamed.mkString(s"The following items are misnamed in $schemaManager:\n", "\n", ""))
       }
 
       var missing = Seq.empty[String]
 
       schemaMetadata.columns foreach { column =>
         if (!schemaMetadata.existsTableItem(column)) {
-          missing :+= s"${tableClassName(column.tableName)}.${column.itemName}"
+          missing :+= s"  ${tableClassName(column.tableName)}.${column.itemName}"
         }
       }
 
       schemaMetadata.slickItems foreach { databaseItem =>
         if (!schemaMetadata.existsSlickMapping(databaseItem)) {
-          missing :+= s"${slickClassName(databaseItem.tableName)}.${databaseItem.itemName}"
+          missing :+= s"  ${slickClassName(databaseItem.tableName)}.${databaseItem.itemName}"
         }
       }
 
-      if (missing.nonEmpty) {
-        fail(missing.mkString(
-          s"Based on the schema in $schemaManager, please ensure that the following tables/columns exist:\n  ",
-          "\n  ",
-          ""))
+      var nullTyped = Seq.empty[String]
+
+      schemaMetadata.columnMetadata foreach { column =>
+        if (column.isNullable == Option(false)) {
+          val jdbcType = JDBCType.valueOf(column.sqlType)
+          if (jdbcType == JDBCType.BLOB || jdbcType == JDBCType.CLOB) {
+            nullTyped :+= s"  $jdbcType column ${column.table.name}.${column.name}"
+          }
+        }
+      }
+
+      if (missing.nonEmpty || misnamed.nonEmpty || nullTyped.nonEmpty) {
+        var failMessage = ""
+
+        if (misnamed.nonEmpty) {
+          failMessage += misnamed.mkString(s"The following items are misnamed in $schemaManager:\n", "\n", "\n")
+        }
+
+        if (missing.nonEmpty) {
+          failMessage += missing.mkString(
+            s"Based on the schema in $schemaManager, please ensure that the following tables/columns exist:\n",
+            "\n", "\n")
+        }
+
+        if (nullTyped.nonEmpty) {
+          failMessage += nullTyped.mkString(
+            s"The following columns should not be nullable due to incompatibilities with MySQL:\n", "\n", "\n")
+        }
+
+        fail(failMessage)
       }
     }
   }
@@ -245,6 +268,75 @@ class ServicesStoreSpec extends FlatSpec with Matchers with ScalaFutures with St
           case _ => Future.successful(())
         }
       } yield ()).futureValue
+    }
+
+    it should "store and retrieve empty clobs" taggedAs DbmsTest in {
+      val workflowUuid = WorkflowId.randomId().toString
+      val callFqn = "call.fqn"
+      val jobIndex = 1
+      val jobAttempt = 1
+      val jobSuccessful = false
+      val jobStoreEntry = JobStoreEntry(workflowUuid, callFqn, jobIndex, jobAttempt, jobSuccessful, None, None, None)
+      val jobStoreSimpletonEntries = Seq(
+        JobStoreSimpletonEntry("empty", "".toClob, "WdlString"),
+        JobStoreSimpletonEntry("aEntry", "a".toClob, "WdlString")
+      )
+      val jobStoreJoins = Seq(JobStoreJoin(jobStoreEntry, jobStoreSimpletonEntries))
+
+      val future = for {
+        _ <- dataAccess.addJobStores(jobStoreJoins)
+        queried <- dataAccess.queryJobStores(workflowUuid, callFqn, jobIndex, jobAttempt)
+        _ = {
+          val jobStoreJoin = queried.get
+          jobStoreJoin.jobStoreEntry.workflowExecutionUuid should be(workflowUuid)
+
+          val emptyEntry = jobStoreJoin.jobStoreSimpletonEntries.find(_.simpletonKey == "empty").get
+          emptyEntry.simpletonValue.toRawString should be("")
+
+          val aEntry = jobStoreJoin.jobStoreSimpletonEntries.find(_.simpletonKey == "aEntry").get
+          aEntry.simpletonValue.toRawString should be("a")
+        }
+        _ <- dataAccess.removeJobStores(Seq(workflowUuid))
+      } yield ()
+      future.futureValue
+    }
+
+    it should "store and retrieve empty blobs" taggedAs DbmsTest in {
+      val testWorkflowState = "Testing"
+
+      val emptyWorkflowUuid = WorkflowId.randomId().toString
+      val emptyWorkflowStoreEntry = WorkflowStoreEntry(emptyWorkflowUuid, "{}".toClob, "{}".toClob, "{}".toClob,
+        testWorkflowState, OffsetDateTime.now.toSystemTimestamp, Option(Array.empty[Byte]).toBlob, "{}".toClob)
+
+      val noneWorkflowUuid = WorkflowId.randomId().toString
+      val noneWorkflowStoreEntry = WorkflowStoreEntry(noneWorkflowUuid, "{}".toClob, "{}".toClob, "{}".toClob,
+        testWorkflowState, OffsetDateTime.now.toSystemTimestamp, None, "{}".toClob)
+
+      val aByte = 'a'.toByte
+      val aByteWorkflowUuid = WorkflowId.randomId().toString
+      val aByteWorkflowStoreEntry = WorkflowStoreEntry(aByteWorkflowUuid, "{}".toClob, "{}".toClob, "{}".toClob,
+        testWorkflowState, OffsetDateTime.now.toSystemTimestamp, Option(Array(aByte)).toBlob, "{}".toClob)
+
+      val workflowStoreEntries = Seq(emptyWorkflowStoreEntry, noneWorkflowStoreEntry, aByteWorkflowStoreEntry)
+
+      val future = for {
+        _ <- dataAccess.addWorkflowStoreEntries(workflowStoreEntries)
+        queried <- dataAccess.queryWorkflowStoreEntries(Int.MaxValue, testWorkflowState, testWorkflowState)
+        _ = {
+          val emptyEntry = queried.find(_.workflowExecutionUuid == emptyWorkflowUuid).get
+          emptyEntry.importsZip.toBytesOption should be(None)
+
+          val noneEntry = queried.find(_.workflowExecutionUuid == noneWorkflowUuid).get
+          noneEntry.importsZip.toBytesOption should be(None)
+
+          val aByteEntry = queried.find(_.workflowExecutionUuid == aByteWorkflowUuid).get
+          aByteEntry.importsZip.toBytesOption.get.toSeq should be(Seq(aByte))
+        }
+        _ <- dataAccess.removeWorkflowStoreEntry(emptyWorkflowUuid)
+        _ <- dataAccess.removeWorkflowStoreEntry(noneWorkflowUuid)
+        _ <- dataAccess.removeWorkflowStoreEntry(aByteWorkflowUuid)
+      } yield ()
+      future.futureValue
     }
 
     it should "close the database" taggedAs DbmsTest in {

